### ISSUE TITLE
[BUG] Reevaluated agents are not matching archive BCs

### ISSUE THREAD

**Issue by aadharna** [icaros-usc/pyribs#109]
- pyribs version: 0.3.1
- Python version: 3.7
- Operating System: Windows
- Emitter style: Improvement

### Description

Trying to do:
So, I am evolving an agent to play Zelda, and as my behavior characteristics, I am using the average action distribution entropy for the agent during the rollout and the number of steps the agent took to achieve the task (get key and get to the door). As the objective, I am +1 if the agents gets the key, +1 if the agent gets the key to the door, -1 if the agent dies to the single monster, and 0 otherwise. Therefore, an objective score of +2 is "solved".

What is going wrong: 
When I extract a solution from the archive at the end of training, e.g: 
![image](https://user-images.githubusercontent.com/16660757/112411683-afd75b00-8cf3-11eb-868b-d39eee38f633.png)

the observed behaviour when I place the agent back into the game does not match the behaviour characterizations that the map says this network weights should achieve. To ensure that I wasn't losing precision due to saving the archive, I have been saving the dataframes in a pkl file. However, that is not a concern here since the archive/dataframe was not saved to disk on this round of experiments.

Now, since the enemy agent moves around randomly, that could make sense that there is variation in the archived BCs and what I'm seeing afterwards. A stochastic environment can defeat many algorithms, but to mitigate noisy signals from the environment, I am averaging my BCs and objective values over 5 rollouts per potential agent. Maybe I just need to do more rollouts per candidate solution.

![image](https://user-images.githubusercontent.com/16660757/112411947-1197c500-8cf4-11eb-8957-0c7a5d3f2b8f.png)

When I run the extracted agent, I am getting entropy values ~1.224 which, for reference, is very far away in my map and due to that higher entropy, the agents are behaving in what looks to be randomly: 

![image](https://user-images.githubusercontent.com/16660757/112412218-8965ef80-8cf4-11eb-80b9-e012282cf133.png)

### Steps to Reproduce

I can send over the full experiment notebook if requested.

**Comment by tehqin** [comment]
Have you tried rerunning the agent twice without saving? Or are your replaying the agent from a save file? If you are saving the archive to a file, how is the archive being saved?

**Comment by tehqin** [comment]
Can you make the environment deterministic somehow and then replay agents? I'm curious if the bug is how the control parameters are being saved (maybe pickle is rounding in a bad way), or if the issue is the variance is too high for the agents (meaning you need to average over more trials).

Can you also try running the agent for 5 additional runs after you eval the agent and see how much the objective and BCs change between evals?

**Comment by aadharna** [comment]
At the moment, I cannot make the enemy agents deterministic. The player agent, however, IS deterministic. 

```
class PolicyNet(nn.Module): 
    def __init__(self, n_actions, depth):
        super(PolicyNet, self).__init__()
        self.conv1 = nn.Conv2d(in_channels=depth, out_channels=8, kernel_size=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(in_channels=8, out_channels=32, kernel_size=1)
        self.fc1 = nn.Linear(128, 48)
        self.fc2 = nn.Linear(48, 24)
        self.fc3 = nn.Linear(24, n_actions)

    def forward(self, x):
        with torch.no_grad():
            x = self.pool(F.relu(self.conv1(x)))
            x = F.relu(self.conv2(x))
            x = torch.flatten(x).unsqueeze(0)
            x = F.relu(self.fc1(x))
            x = F.relu(self.fc2(x))
            x = F.softmax(self.fc3(x), dim=1).squeeze()
            _, action = torch.max(x, 0)
        
            # break data out of tensor
            # action = predicted.data.numpy()
            action_dist = Categorical(x)
            # action = action_dist.sample()

        return action.numpy(), -action_dist.log_prob(action), action_dist.entropy().mean(), 0
```

Regarding saving: 

```
import pickle
def save_obj(obj, folder, name):
    path = os.path.join(folder, name) + '.pkl'
    if os.path.exists(path):
        os.remove(path)
    with open(path, 'wb+') as f:
        pickle.dump(obj, f, pickle.HIGHEST_PROTOCOL)
        
def load_obj(folder, name):
    path = os.path.join(folder, name)
    try:
        with open(path, 'rb') as f:
            f = pickle.load(f)
    except EOFError:
        f = {"resend":True}
    return f
```
However, the weights are being extracted directly from the dataframe which was created by the `archive.to_pandas()` call. I am testing this explicitly without saving the weights to disk since you and I am mentioned that saving the csv via pandas could screw things up and now on an earlier run when I did save the weights using the above, the behaviour was still not matching the archive bin. 

> Can you also try running the agent for 5 additional runs after you eval the agent and see how much the objective and BCs change between evals?

Here's the extracted agent for 10 additional runs. The order here is: [objective, bc1, bc2]
![image](https://user-images.githubusercontent.com/16660757/112415411-26775700-8cfa-11eb-9d5b-8d1d9bc0675d.png)

And for reference, the bin I grabbed was 40522: 
![image](https://user-images.githubusercontent.com/16660757/112415720-b3baab80-8cfa-11eb-97ed-218629baec5f.png)

----

I'll set an experiment to run tonight where I evaluate the weights as I'm doing right now (5 runs, averaged), and then I'll do another 5 before continuing on to the next candidate solution and report back tomorrow.

**Comment by btjanaka** [comment]
Hi @aadharna, thanks for submitting the issue. For saving the dataframe to a pickle file and loading the dataframe from a pickle file, can you use [pandas.DataFrame.to_pickle](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_pickle.html) and [pandas.read_pickle](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_pickle.html), respectively?

For your initial question, I do not understand how the heatmap shows the random behavior. The entropy value of 1.224 seems well within the values in the heatmap?

Finally, how are you extracting weights from the network and loading weights into the network?

**Comment by aadharna** [comment]
Regarding how I am unloading/loading weights into the network: 

```
class PyTorchObjective():
    """PyTorch objective function, wrapped to be called by optimizer."""

    def __init__(self, agent):
        self.agent = agent
        # make an x0 from the parameters in this module
        parameters = OrderedDict(agent.named_parameters())
        self.param_shapes = {n: parameters[n].size() for n in parameters}
        # ravel and concatenate all parameters to make x0
        self.x0 = np.concatenate([parameters[n].data.numpy().ravel()
                                  for n in parameters])
            
    def unpack_parameters(self, x):
        """will be supplied a 1D array, chop it up for each parameter."""
        i = 0
        named_parameters = OrderedDict()
        for n in self.param_shapes:
            param_len = reduce(lambda x, y: x * y, self.param_shapes[n])
            # slice out a section of this length
            param = x[i:i + param_len]
            # reshape according to this size, and cast to torch
            param = param.reshape(*self.param_shapes[n])
            named_parameters[n] = torch.from_numpy(param)
            # update index
            i += param_len
        return named_parameters
    
    def update_nn(self, weights):
        state_dict = self.unpack_parameters(weights)
        self.agent.load_state_dict(state_dict)
 ```

----  
Regarding the entropy values observed on repeat evaluation: 

The value is a valid one according to the map, yes. But the weights I extracted said they came from an area of the map indexed by (entropy, steps) = (0.002, 0.08). The behavior of a policy with that level of action-entropy and number of steps used should be basically a direct path from the agent's starting position to the key then goal. (The steps number is normalized to between 0 and 1.) However, when I reran the weights in the environment I received a tuple of (1.23, 0.78)

----  
I originally used pandas' to_pickle but got some error on read (I don't recall what it was because then I just decided to pickle the entire object). But when the script frees back up (there are about 2.5 hours left on the script according to tqdm) I'll try the pandas to/from pickling again and report back.

**Comment by btjanaka** [comment]
The weight loading / unloading looks correct, though it may depend on whether `named_parameters` always returns parameters in the same order.

Do let me know what happens with the pandas pickling. The dataframe should only contain floats and integers, so pickling should work as both of those are pickle-able. Furthermore, saving to CSV should work because there are only floats and integers.

**Comment by aadharna** [comment]
> Do let me know what happens with the pandas pickling.  

I did the pickling through pandas and the saving/loading worked out fine. I guess my brain was conflating errors. 

The saved-via-pandas-pickling weights still produced behaviours that didn't match what the archive cell said it should be though. Now that that experiment is done, I launched the next one where I remove the enemy and see if the weights in that scheme will fix the issue (but my guess is it won't because the agent only sees a very limited part of the environment at any given time and that part is not guaranteed to contain the enemy).

**Comment by btjanaka** [comment]
Are you comfortable with sending over the notebook? I can try looking through it for the bug. My email is bryon@btjanaka.net. A private GitHub Gist or repo will also work.

**Comment by aadharna** [comment]
Sent. Don't hesitate to ping me if I forgot something you need and/or you have questions.

**Comment by btjanaka** [comment]
Resolved offline. It looks like pyribs itself is fine, and the issue seems to be with the environment evaluations.
